crud/common/sharding_key.lua

```diff
@@ -0,0 +1,229 @@
+-- it contains only parts key with fieldno's.
+local function as_index_object(space_name, space_format, sharding_key_def)
+    dev_checks('string', 'table', 'table')
+
+    local fieldnos = {}
+    local fieldno_map = utils.get_format_fieldno_map(space_format)
+    for _, field_name in ipairs(sharding_key_def) do
+        local fieldno = fieldno_map[field_name]
+        if fieldno == nil then
+            return nil, WrongShardingConfigurationError:new(
+                "No such field (%s) in a space format (%s)", field_name, space_name)
+        end
+        table.insert(fieldnos, {fieldno = fieldno})
+    end
+
+    return {parts = fieldnos}
+end
+
+-- Return a map or nil when metadata is not available.
+function sharding_key_module.fetch_on_storage()
+    local sharding_key_space = box.space._ddl_sharding_key
+    if sharding_key_space == nil then
+        return nil
+    end
+
+    local SPACE_NAME_FIELDNO = 1
+    local SPACE_SHARDING_KEY_FIELDNO = 2
+    local metadata_map = {}
+    for _, tuple in sharding_key_space:pairs() do
+        local space_name = tuple[SPACE_NAME_FIELDNO]
+        local sharding_key_def = tuple[SPACE_SHARDING_KEY_FIELDNO]
+        local space_format = box.space[space_name]:format()
+        metadata_map[space_name] = {
+            sharding_key_def = sharding_key_def,
+            space_format = space_format,
+        }
+    end
+
+    return metadata_map
+end
+
+local sharding_key_fetch_in_progress = false
+local sharding_key_fetch_cond = fiber.cond()
+
+-- Under high load we may get a case when many fibers will fetch metadata from
+-- storages. It is not good from performance point of view. reentrant_fetch()
+-- wraps a fetch_on_storage() to limit a number of fibers that fetches a
+-- metadata by a single one, other fibers will wait for a
+-- FETCH_SHARDING_KEY_TIMEOUT and then make one more attempt to fetch metadata
+-- if a cache is still empty.
+local function reentrant_fetch()
+    if cache.sharding_key_as_index_obj_map ~= nil then
+        return cache.sharding_key_as_index_obj_map
+    end
+
+    local timeout_deadline = fiber.clock() + const.FETCH_SHARDING_KEY_TIMEOUT
+    local timeout_is_exceeded
+    while sharding_key_fetch_in_progress == true do
+        timeout_is_exceeded = false
+        local timeout = timeout_deadline - fiber.clock()
+        if sharding_key_fetch_cond:wait(timeout) then
+            timeout_is_exceeded = true
+        end
+        if cache.sharding_key_as_index_obj_map ~= nil then
+            return cache.sharding_key_as_index_obj_map
+        end
+        if timeout_is_exceeded == true then
+            return nil, FetchShardingKeyError:new(
+                            'Waiting for sharding key to be fetched is timed out (%d sec)',
+                            const.FETCH_SHARDING_KEY_TIMEOUT)
+        end
+    end
+
+    sharding_key_fetch_in_progress = true -- Start of critical section.
+    local metadata_map, err = call.any(FETCH_FUNC_NAME, {}, {
```

Comment:
I met the following situation while solving #213 and running unit tests
![image](https://user-images.githubusercontent.com/20455996/143231769-a4fb1f8f-f529-4ddc-a094-26bfbd6d3fd0.png)
Unconfigured `vshard.router` throws error in `call.any`, critical section is broken and sharding key module is unusable, Maybe it should be processed more carefully.

Reply:
> I met the following situation while solving #213 and running unit tests image
> Unconfigured vshard.router throws error in call.any, critical section is broken and sharding key module is unusable. 

Do you have a reproducer?

Reply:
#242 without second commit

Reply:
Thanks for catching this! I made changes and after rebasing your branch on top of my branch the problem is gone.